pypesto/objective/finite_difference.py

```diff
@@ -1,13 +1,215 @@
+
+        Parameters
+        ----------
+        x:
+            Current parameter vector, shape (n_par,).
+        fval:
+            fun(x), to avoid re-evaluation. Scalar- or vector-valued.
+        fun:
+            Function whose 1st-order derivative to approximate.
+            Scalar- or vector-valued.
+        fd_method:
+            FD method employed by
+            :class:`pypesto.objective.finite_difference.FD`, see there.
+        """
+        # scalar to vector
+        if isinstance(self.delta, float):
+            self.delta: np.ndarray = self.delta * np.ones(shape=x.shape)
+
+        self.steps += 1
+
+        # return if no update needed
+        if self.delta is not None:
+            if (
+                self.update_condition == FDDelta.SPACE and
+                np.sum((x - self.x0)**2) <= 0.5 * np.sqrt(len(x))
+            ):
+                return
+            if (
+                self.update_condition == FDDelta.STEPS and
+                (self.steps-1) % self.max_steps != 0
+            ):
+                return
+
+        # update reference point
+        self.x0 = x
+        # actually update
+        self._update(x=x, fval=fval, fun=fun, fd_method=fd_method)
+
+        if self.delta.shape != x.shape:
+            # this should not happen
+            raise ValueError("Shape mismatch.")
+
+    def _update(
+        self,
+        x: np.ndarray,
+        fval: Union[float, np.ndarray],
+        fun: Callable,
+        fd_method: str,
+    ) -> None:
+        """Actually update. Wants to be called in `update` explicitly.
+
+        Run FDs with various deltas and pick the ones, separately for each
+        parameter, with the best stability properties.
+
+        The parameters are the same as for
+        :func:`pypesto.objective.finite_difference.FDDelta.update`.
+        """
+
+        # calculate gradients for all deltas for all parameters
+        nablas = []
+        # iterate over deltas
+        for delta in self.test_deltas:
+            # calculate Jacobian with step size delta
+            delta_vec = delta * np.ones_like(x)
+            nabla = fd_nabla_1(
+                x=x, fval=fval, f_fval=fun, delta_vec=delta_vec,
+                fd_method=fd_method)
+
+            nablas.append(nabla)
+
+        # shape (n_delta, n_par, ...)
+        nablas = np.array(nablas)
+
+        # The stability vector is the the absolute difference of Jacobian
+        #  entries towards smaller and larger deltas, thus indicating the
+        #  change in the approximation when changing delta.
+        # This is done separately for each parameter. Then, for each the delta
+        #  with the minimal entry and thus the most stable behavior
+        #  is selected.
+        stab_vec = np.full(shape=nablas.shape, fill_value=np.nan)
+        stab_vec[1:-1] = 0.5 * (
+            np.abs(nablas[2:] - nablas[1:-1]) +
+            np.abs(nablas[1:-1] - nablas[:-2])
+        )
```

Comment:
```
stab_vec[1:-1] = np.mean(np.abs(
    [nablas[2:] - nablas[1:-1], 
     nablas[1:-1] - nablas[:-2]]
))
```
?

Reply:
Should hopefully do the same, and makes it clearer what actually happens

Reply:
alternatively, you add a nan-vector in the first and last position and just use np.nanmean over the whole thing, whic avoids handling the corner cases separately